Update design doc core part (#2741)
open-edge-platform/training_extensions: for_developers/product_design.md

# Product Design

## Our Product Mission

1. **Provide a deep learning model training framework that encompasses various tasks solvable by deep learning models.** Our product aims to be convenient for developers, allowing them to easily expand support for tasks and models.

2. **Empower users to seamlessly and intuitively navigate the OpenVINO™-based workflow, from model training to inference.** To ensure a user-friendly experience throughout, our product should have user-friendly CLI and Python API entrypoints.

3. **Support Intel devices for accelerating deep learning model training:** We offer best practices for training from scratch or fine-tuning popular model architectures using Intel devices.

## Product Design Principles

To realize our product missions, we establish the following design principles and details, providing a comprehensive overview of our product's structure.

### Core

1. **Static Typing System for Task-Data-Model to Accelerate Task and Model Development Cycles**

   _"Advocates of static typing argue that ... a better design time developer experience..."_[^1]

   Our framework should support various deep learning tasks simultaneously.

>    To achieve this mission, it is essential to make it developer-friendly for expanding our product in both task and model dimensions.

Reviewer: please refer to data dimension here

Author: ```suggestion
   To achieve this mission, it is essential to make it developer-friendly for expanding our product across three dimensions: task, data, and model.
   The key insight here is that, while these dimensions may seem independent, they can be interconnected into a single dimension, specifically the task, where both the data and model depend on.
```
How about this?

Reviewer: looks good to me.

   We believe that nice abstractions with static typing can help developers in this regard.

>    We introduced the **Task-Data-Model** abstraction with static typing.

Reviewer: How about `Therefore, we here introduce the **Task-Data-Model** ...`?

Author: ```suggestion
   Therefore, we here introduce the **Task-Data-Model** abstraction with static typing.
```

   4. `OTXModel`: PyTorch module (`nn.Module`) defined for each `OTXTaskType`.
      Its `forward()` function has strict typing: 1) `OTXBatchDataEntity` as inputs and 2) `OTXBatchLossEntity` or `OTXBatchPredEntity` as outputs.

   On the other hand, we can see embodiments from the abstractions (blue-colored boxes) for the detection task (`OTXTaskType.DETECTION`):

   1. `DetDataEntity` and `DetBatchDataEntity`: Python dataclass objects including `bboxes` and `labels` fields for the detection task.
      `DetBatchDataEntity` is used for the input of `OTXDetectionModel`'s `forward()` function.
   2. `DetBatchPredEntity`: Python dataclass object including the `scores` field beyond `DetBatchDataEntity`.
      It is used for the output of `OTXDetectionModel`'s `forward()` function at inference mode.

   As observed, a developer aiming to add a new model to OTX for the detection task only needs to consider the strictly typed input and output (`DetBatchDataEntity` and `DetBatchPredEntity`).

   Conversely, in the event of a new task requirement, we can incorporate it by implementing embodiments from the base, such as `OTXTaskType`, `OTXDataEntity`, `OTXBatchDataEntity`, and `OTXModel`.

> 2. **Single Engine but Capable of Incorporating Multiple Popular Model Training Frameworks**

Reviewer: How about `Other Public Model Training Frameworks` instead of `Multiple Popular`?

Author: ```suggestion
2. **Single Engine but Capable of Incorporating Other Public Model Training Frameworks**
```

   _"Don't Reinvent the Wheel"_[^2]

>    We chose PyTorch Lightning[^3] as the primary engine for our training framework.

Reviewer: `chose` -> `choose`. I think it is better to keep using same verb tense in overall document.

Author: ```suggestion
   We choose PyTorch Lightning[^3] as the primary engine for our training framework.
```

   To simplify and enhance this process in the model development lifecycle, we introduce Datumaro[^7], a member of the OpenVINO™ open-source ecosystem.
   Datumaro allows the import of various data formats from around the world.
   Please refer to the following Figure.

   | ![Support Various Data Format](./images/product_design/support_various_data_format.png) |
   | :-------------------------------------------------------------------------------------: |
   | Figure 3. Only the root directory path is provided by the user to make a data pipeline  |

   In Figure 3, a popular data format such as COCO, VOC, YOLO, or Cityscape can be imported with Datumaro.
   While Figure 3 illustrates the entire data pipeline, the only input required from the user is `data_root`.
   This simplicity allows users to enjoy a convenient model training workflow without worrying about the input data pipeline.

   Another advantage highlighted in Figure 3 is the support for multiple data augmentation pipelines.
   The default recommendation is TorchVision[^8].
   However, as shown, there are other pipelines such as MMPretrain, MMDetection, etc.

>    This flexibility enables developers to freely compose the model and the data augmentation pipeline from various frameworks, accelerating the creation of custom model training templates for specific problem domains.

Reviewer: Sorry that I felt these are not sufficient. Why don't you adding some backgrounds of introduction to factory type transformlib to support various customized augmentations per frameworks?

Author: Updated in https://github.com/openvinotoolkit/training_extensions/pull/2741/commits/3051270768fca55c33a0009368687c6a2cf80fb3

Authors: @wonjuleee @vinnamkim

### Entrypoint

TBD @samet-akcay @harimkang

### Intel Device Support

TBD

[^1]: Meijer, Erik, and Peter Drayton. "Static typing where possible, dynamic typing when needed: The end of the cold war between programming languages." OOPSLA, 2004.
[^2]: https://en.wikipedia.org/wiki/Reinventing_the_wheel#In_software_development
[^3]: https://lightning.ai/pytorch-lightning
[^4]: https://github.com/open-mmlab/mmdetection
[^5]: https://github.com/open-mmlab/mmdetection/tree/main/configs/rtmdet
[^6]: https://www.netapp.com/media/16928-wp-7299.pdf
[^7]: https://github.com/openvinotoolkit/datumaro

> [^8]: https://pytorch.org/vision/0.16/transforms.html

Reviewer: links may not be working correctly now. Could you check this again?

Reviewer: Sorry for my misunderstanding.

   Let's consider an example of importing a model from MMDetection[^4]:

   | ![Reuse Model](./images/product_design/reuse_model.png) |
   | :-----------------------------------------------------: |
   | Figure 2. Import RTMDetTiny[^5] model from MMDetection  |

   In Figure 2, `MMDetCompatibleModel` implements `create_model()` to generate the `RTMDetTiny` model and registers it as a child module.
   At this point, the `RTMDetTiny` model requires `DataSample` input from MMDetection and produces a Python dictionary for its output.
   These data entity types do not comply with our framework.
   To resolve this issue, `MMDetCompatibleModel` implements `customize_inputs()` and `customize_outputs()` functions to convert them into our format.
   With this design, we can explicitly support major frameworks. We plan to continually expand the supported frameworks using this approach.

3. **Support Various Data Formats without Worrying about Customizing the Input Data Pipeline**

>    _Much of the time for an AI project is likely to be spent on data-related tasks ..._[^6]

Author: ```suggestion
   _"Much of the time for an AI project is likely to be spent on data-related tasks ..."_[^6]
```
